File: src/infinicore/ops/sddmm/sddmm.cc
#include "infinicore/ops/sddmm.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(SDDMM);

SDDMM::SDDMM(Tensor c_values, SpMat c, const Tensor &a, const Tensor &b, float alpha, float beta) {
    INFINICORE_ASSERT(c);
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(c_values, c->values(), c->crow_indices(), c->col_indices(), a, b);
    INFINICORE_GRAPH_OP_DISPATCH(c->device().getType(), c_values, c, a, b, alpha, beta);
}

void SDDMM::execute(Tensor c_values, SpMat c, const Tensor &a, const Tensor &b, float alpha, float beta) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(SDDMM, c_values, c, a, b, alpha, beta);
}

void sddmm_(SpMat c, const Tensor &a, const Tensor &b, float alpha, float beta) {
    SDDMM::execute(c->values(), c, a, b, alpha, beta);
}

} // namespace infinicore::op

File: python/infinicore/ops/sddmm.py
from infinicore.lib import _infinicore
from infinicore.spmat import SpMat
from infinicore.tensor import Tensor


def sddmm(c: SpMat, a: Tensor, b: Tensor, *, alpha=1.0, beta=0.0):
    if not isinstance(c, SpMat):
        raise TypeError("sddmm expects a CSR SpMat as the sampled output")

    _infinicore.sddmm_(c._underlying, a._underlying, b._underlying, alpha, beta)
    return c
